syntax = "proto3";

import "attr.proto";

package pbuf;

service BiobtreeService {
  rpc Search        (SearchRequest)         returns (SearchResponse);
  rpc Mapping       (MappingRequest)        returns (MappingResponse);
  rpc Entry         (EntryRequest)          returns (EntryResponse);
  rpc Page          (PageRequest)           returns (PageResponse);
  rpc Filter        (FilterRequest)         returns (FilterResponse);
  rpc Meta          (MetaRequest)           returns (MetaResponse);
  rpc ListGenomes   (ListGenomesRequest)    returns (ListGenomesResponse);
}

message SearchRequest {
  	repeated string terms=1;
    string dataset=2;
    string page=3;
    string query=4;
    bool detail=5;
    bool url=6;
}

message SearchResponse {
  	Result results =1;
}

message MappingRequest {
  	repeated string terms = 1;
    string dataset=2;
    string query = 3;
    string page =4;
}

message MappingResponse {
    MapFilterResult results =1;
}

message EntryRequest {
  	string identifier = 1;
    string dataset=2;
}

message EntryResponse {
  	Xref result =1;
}

message PageRequest {
  	string identifier = 1;
    string dataset=2;
    int32 page =3;
    int32 total=4;
}

message PageResponse {
  	Result result =1;
}

message FilterRequest {
  	string identifier = 1;
    string dataset=2;
    repeated string filters =3;
    int32 page =4;
}

message FilterResponse {
  	Result result =1;
}

message MetaRequest {
}

message MetaResponse {
   map<string, MetaKeyValue> results = 1;
}

message MetaKeyValue{
  map<string, string> keyvalues = 1;
}

message ListGenomesRequest {
  	string type = 1;
}

message ListGenomesResponse {
    string results = 1;
}

message Result {
  	repeated Xref results = 1;
    string nextpage=2;
}

message MapFilterResult {
  	repeated MapFilter results = 1;
    string nextpage=2; 
}

message MapFilter {
  	 Xref source = 1;
     repeated Xref targets = 2;
}

message Xref {

  uint32 dataset = 1;
  bool is_link=2;
  string identifier=3;// transient
  string keyword =4;  // transient

  oneof attributes{

    bool empty = 5 ;
    UniprotAttr uniprot=6;
    OntologyAttr ontology=7;
    HgncAttr hgnc= 8;
    InterproAttr interpro= 9;
    EnsemblAttr ensembl= 10;
    TaxoAttr taxonomy= 11;
    HmdbAttr hmdb= 12;
    EnaAttr ena= 13;
    PdbAttr pdb= 14;
    DrugbankAttr drugbank= 15;
    OrphanetAttr orphanet= 16;
    ReactomeAttr reactome= 17;
    ChemblAttr chembl=18;
    UniprotFeatureAttr ufeature=19;

  }

  uint32 count =20;
  repeated XrefEntry entries=21;
  repeated XrefDomainCount dataset_counts=22;
  map<uint32,PageInfo> datasetPages=23;
  repeated string pages=24;
  string url =25; // transient
}

message PageInfo {
  repeated string pages=1;
}

message XrefEntry {

  uint32 dataset = 1;
  string identifier = 2;

}

message XrefDomainCount {

  uint32 dataset = 1;
  uint32 count =2;

}

message Alias {
    repeated string identifiers = 1;
}